Mathlib/LinearAlgebra/FreeModule/Int.lean

```diff
@@ -0,0 +1,157 @@
+/-
+Copyright (c) 2024 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import Mathlib.Data.ZMod.Quotient
+import Mathlib.GroupTheory.Index
+import Mathlib.LinearAlgebra.FreeModule.PID
+
+/-! # Index of submodules of free ℤ-modules (considered as an `AddSubgroup`).
+
+This file provides lemmas about when a submodule of a free ℤ-module is a subgroup of finite
+index.
+
+-/
+
+
+namespace Basis.SmithNormalForm
+
+lemma toAddSubgroup_index_ne_zero_iff {M : Type*} {m n : ℕ} [AddCommGroup M] {N : Submodule ℤ M}
```

Comment:
Can you add a docstring? More precisely, I think it is worth to mention that the theorem is true because if `N` has infinite index, then `N.index` is `0` in Lean (I was puzzled by this for 2 minutes before realizing what the theorem was saying).

Also, can you replace `Fin m` by a generic type `ι` (maybe with `Fintype`) and use a cardinality statement?

Reply:
I've added a doc string and changed everything to use a generic `Fintype` (or `Finite` type when `Fintype.card` isn't used in the lemma statements).

Reply:
Perfect, thanks!